src/osp/Active/SysRender.h

```diff
@@ -0,0 +1,164 @@
+#include <Magnum/GL/Renderbuffer.h>
+
+namespace osp::active
+{
+
+/* ########## USER COMPONENTS ########## */
+
+/*
+ * Stores a mesh to be drawn by the renderer
+ */
+struct ACompMesh
+{
+    osp::DependRes<Magnum::GL::Mesh> m_mesh;
+};
+
+/*
+ * Diffuse texture component
+ */
+struct ACompDiffuseTex
+{
+    osp::DependRes<Magnum::GL::Texture2D> m_tex;
+};
+
+/*
+ * A surface that can be rendered to
+ */
+struct ACompRenderTarget
+{
+    Magnum::Vector2i m_size;
+    osp::DependRes<Magnum::GL::Framebuffer> m_fbo;
+};
+
+/*
+ * Represents an entity which requests the scene to be drawn to m_target
+ */
+struct ACompRenderingAgent
+{
+    ActiveEnt m_target;  // Must have ACompRenderTarget
+};
+
+/*
+ * Describes the type of view used by a rendering agent
+ */
+struct ACompPerspective3DView
+{
+    ActiveEnt m_camera;
+};
+
+/* An object that is completely opaque */
+struct ACompOpaque {};
+
+/* An object with transparency */
+struct ACompTransparent {};
+
+/* Visibility state of this object */
+struct ACompVisible {};
+
+/* ########## RENDER INTERNAL COMPONENTS ########## */
+
+/*
+ * Stores a shader draw function to be used to draw the object
+ */
+struct ACompShader
+{
+    ShaderDrawFnc_t m_drawCall;
+};
+
+struct ACompRenderer
+{
+    std::string m_name;
+};
+
+struct ACompFBOColorAttachment
+{
+    DependRes<Magnum::GL::Texture2D> m_tex;
+};
+
+struct RenderPipeline
+{
+    RenderOrder_t m_order;
+};
+
+class SysRender
+{
+public:
+    /* Register update and render functions to scene */
+    static void add_functions(ActiveScene& rScene);
+
+    /* Configure rendering passes, init runtime resources */
+    static void setup(ActiveScene& rScene);
+
+    /* Update scene components, manage drawable objects */
+    static void update(ActiveScene& rScene);
+
+    /* Execute render passes */
+    static void render(ActiveScene& rScene, ACompCamera const& camera);
+
+    static ActiveEnt get_default_rendertarget(ActiveScene& rScene)
+    { return rScene.hier_get_root(); }
+
+    static void display_default_rendertarget(ActiveScene& rScene);
+private:
+
+    /* Temporary?: define render passes */
+    static RenderPipeline create_render_pipeline();
+
+    /*
+     * Draw the specified list of objects from the point of view of a camera
+     * 
+     * @param rScene - The active scene containing the data
+     * @param drawlist - The EnTT view containing a list of drawable objects
+     *                   which possess an ACompShader component
+     * @param camera - The camera whose point of view to render from
+     */
+    template <typename VIEW_T>
+    static void draw_group(ActiveScene& rScene,
+        VIEW_T& drawlist, ACompCamera const& camera);
+};
+
+template<typename VIEW_T>
+void SysRender::draw_group(ActiveScene& rScene, VIEW_T& drawlist, ACompCamera const& camera)
+{
+    for (ActiveEnt e : drawlist)
+    {
+        auto shader = drawlist.template get<ACompShader>(e);
+
+        shader.m_drawCall(e, rScene, camera);
+    }
+}
```

Comment:
For stuff like filling a g-buffer or common objects, there should be a "draw group using this shader only" function too.

Reply:
That would usually be handled as one of the render passes. Can you provide an example?